src/opf/ed.jl

```diff
@@ -0,0 +1,467 @@
+using SparseArrays
+
+mutable struct EconomicDispatch <: PM.AbstractPowerModel end
+
+function build_opf(::Type{EconomicDispatch}, data::Dict{String,Any}, optimizer;
+    T=Float64,
+    soft_power_balance::Bool=false,
+    soft_reserve_requirement::Bool=false,
+    soft_thermal_limit::Bool=false,
+    power_balance_penalty=350000.0,
+    reserve_shortage_penalty=110000.0,
+    thermal_penalty=150000.0,
+    iterative_ptdf::Bool=true,
+    iterative_ptdf_tol=1e-6,
+    max_ptdf_iterations=10,
+    max_ptdf_per_iteration=5,
+)
+    # Cleanup and pre-process data
+    PM.standardize_cost_terms!(data, order=2)
+    PM.calc_thermal_limits!(data)
+    ref = PM.build_ref(data)[:it][:pm][:nw][0]
+
+    # Check that slack bus is unique
+    length(ref[:ref_buses]) == 1 || error("Data has $(length(ref[:ref_buses])) slack buses (expected 1).")
+    i0 = first(keys(ref[:ref_buses]))
+
+    power_balance_penalty >= 0.0 || error("EconomicDispatch option power_balance_penalty must be non-negative")
+    reserve_shortage_penalty >= 0.0 || error("EconomicDispatch option reserve_shortage_penalty must be non-negative")
+    thermal_penalty >= 0.0 || error("EconomicDispatch option transmission_penalty must be non-negative")
+    max_ptdf_iterations > 0 || error("EconomicDispatch option max_ptdf_iterations must be a positive integer")
+    max_ptdf_per_iteration > 0 || error("EconomicDispatch option max_ptdf_per_iteration must be a positive integer")
+
+    # Grab some data
+    N = length(ref[:bus])
+    G = length(ref[:gen])
+    E = length(data["branch"])
+    L = length(ref[:load])
+
+    pmin = [ref[:gen][g]["pmin"] for g in 1:G]
+    pmax = [ref[:gen][g]["pmax"] for g in 1:G]
+    PD = sum(ref[:load][l]["pd"] for l in 1:L)
+    pfmax = [data["branch"]["$e"]["rate_a"] for e in 1:E]
+
+    minimum_reserve = get(data, "minimum_reserve", 0.0)
+    rmin = (minimum_reserve > 0.0) ? [ref[:gen][g]["rmin"] for g in 1:G] : zeros(T, G)
+    rmax = (minimum_reserve > 0.0) ? [ref[:gen][g]["rmax"] for g in 1:G] : zeros(T, G)
+
+    model = JuMP.GenericModel{T}(optimizer)
+    model.ext[:opf_model] = EconomicDispatch  # for internal checks
+    model.ext[:solve_metadata] = Dict{Symbol,Any}(
+        :iterative_ptdf => iterative_ptdf,
+        :iterative_ptdf_tol => iterative_ptdf_tol,
+        :max_ptdf_iterations => max_ptdf_iterations,
+        :max_ptdf_per_iteration => max_ptdf_per_iteration,
+    )
+
+    # Variables
+
+    JuMP.@variable(model, pmin[g] <= pg[g in 1:G] <= pmax[g])
+    JuMP.@variable(model, rmin[g] <= r[g in 1:G] <= rmax[g])
+    JuMP.@variable(model, pf[e in 1:E])
+
+    # for upper bounds, see https://github.com/AI4OPT/OPFGenerator/pull/64/files#r1529320645
```

Comment:
Upper bounds on which variables?

Reply:
Looks like this link is broken.. this used to point to a comment you made regarding how to compute upper bounds for the slack variables